agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package datadog.trace.bootstrap
+
+import io.opentelemetry.auto.bootstrap.InternalJarURLHandler
+import io.opentelemetry.auto.util.test.AgentSpecification
+import spock.lang.Shared
+
+class InternalJarURLHandlerTest extends AgentSpecification {
+
+  @Shared
+  URL testJarLocation = new File("src/test/resources/classloader-test-jar/testjar-jdk8").toURI().toURL()
+
+
+  def "test extract packages"() {
+    setup:
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    expect:
+    packages == handler.getPackages()
+
+    where:
+    dir      | packages
+    "parent" | ['a', 'a.b', 'a.b.c'].toSet()
+    "child"  | ['x', 'x.y', 'x.y.z'].toSet()
+  }
+
+  def "test get URL"() {
+    setup:
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    when:
+    URLConnection connection = handler.openConnection(new URL('file://' + file))
+    assert connection != null
+    byte[] data = new byte[128]
+    int read = connection.getInputStream().read(data)
+    then:
+    read > 0
+
+    where:
+    dir      | file
+    "parent" | '/a/A.class'
+    "parent" | '/a/b/B.class'
+    "parent" | '/a/b/c/C.class'
+    "child"  | '/x/X.class'
+    "child"  | '/x/y/Y.class'
+    "child"  | '/x/y/z/Z.class'
+  }
+
+
+  def "test read class twice"() {
+    // guards against caching the stream
+    setup:
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    when:
+    URLConnection connection = handler.openConnection(new URL('file://' + file))
+    assert connection != null
+    InputStream is = connection.getInputStream()
+    is.close()
+    connection = handler.openConnection(new URL('file://' + file))
+    assert connection != null
+    is = connection.getInputStream()
+    byte[] data = new byte[128]
+    int read = is.read(data)
+
+    then:
+    read > 0
+
+    where:
+    dir      | file
+    "parent" | '/a/A.class'
+    "parent" | '/a/b/B.class'
+    "parent" | '/a/b/c/C.class'
+    "child"  | '/x/X.class'
+    "child"  | '/x/y/Y.class'
+    "child"  | '/x/y/z/Z.class'
+  }
+
+  def "handle not found"() {
+    setup:
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    when:
+    handler.openConnection(new File(file).toURI().toURL())
+    then:
+    // not going to specify (and thereby constrain) the sub type because it doesn't matter
+    thrown IOException
+
+    // permuted
+    where:
+    dir      | file
```

Comment:
Why do you want to test several non-existent paths here? Why 1 is not enough?

Reply:
updated